### ISSUE TITLE
Some uboot/ubnt stuff I have worked on in the past.

### ISSUE THREAD

**Issue by andrewhodel** [c-mauderer/openwrt-Ubiquity-Aircube-notes#1]
I think this may help you, from reading your notes here

https://github.com/c-mauderer/openwrt-Ubiquity-Aircube-notes/blob/master/notes/Notes.asciidoc

## 1

```
You can downgrade uboot once you get in with the serial console, `urescue -f -e`.
```

## 2

When trying to use a new version of AirOS or openwrt to reflash an old version of AirOS just use the old AirOS binary flash tool (THAT IS ALREADY INSTALLED ON THE AIROS FIRMWARE)

The issue is that they are releasing new firmwares with different flash layouts and basically the uboot tftp loaded as well as the web ui won't allow LEDE or OpenWRT images to be flashed.

As we know the devices are the same hardware, which means the same binaries will run.

So, regardless of the AirOS version we can scp the binLEDEary which is the updater and then flash our openwrt image.

```
binwalk -e XW.v5.6.15.30572.170328.1052.bin
scp _XW.v5.6.15.30572.170328.1052.bin.extracted/squashfs-root/bin/ubntbox ubnt@10.103.68.69:/tmp/
scp XW.v5.6.15.30572.170328.1052.bin ubnt@10.103.68.69:/tmp/fwupdate.bin
ssh ubnt@10.103.68.69
cd /tmp; /tmp/ubntbox fwupdate.real -m fwupdate.bin
```

This is important because the older AIROS versions didn't expect signing in uboot.

## 3

You can install `mtd` on AIROS by taking the executable from an older version of AIROS (dissect the binary) if `mtd` is not already on the version of AIROS currently on the device.  Make sure the AIROS build is for ath79 not ath79.

first ssh ubnt@192.168.1.20

check if mtd exists, if not, download the 3.7.58.6385 firmware that's here in mindmap so it can be downgraded to include mtd.

then

```
scp BZ.......bin ubnt@192.168.1.20:/tmp/
ssh ubnt@192.168.1.20
cd /tmp
fwupdate.real -m BZ.....bin
```

### once you have mtd in airos

```
BZ.v3.7.40# mtd write /tmp/openwrt-ar71xx-generic-ubnt-unifiac-lite-squashfs-sysupgrade.bin kernel0
Unlocking kernel0 ...
Erasing kernel0 ...
Writing from /tmp/openwrt-ar71xx-generic-ubnt-unifiac-lite-squashfs-sysupgrade.bin to kernel0 ...  [e/w]

BZ.v3.7.40# mtd erase kernel1
Unlocking kernel1 ...
Erasing kernel1 ...
```

Then write a nullbyte to the bs partition (check /proc/mtd for the correct path) to make it boot from kernel0, which is the partition that sysupgrade writes to.

```
BZ.v3.9.19# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00060000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00790000 00010000 "kernel0"
mtd3: 00790000 00010000 "kernel1"
mtd4: 00020000 00010000 "bs"
mtd5: 00040000 00010000 "cfg"
mtd6: 00010000 00010000 "EEPROM"

BZ.v3.9.19# dd if=/dev/zero bs=1 count=1 of=/dev/mtd4
1+0 records in
1+0 records out
1 byte copied, 7.3437e-05 s, 13.6 kB/s
```

## 4

Beyond that, you can always compile static linked binaries in the openwrt buildroot then copy them to the AIROS image to flash using the MTD ioctls.

**Comment by c-mauderer** [comment]
Hello @andrewhodel,

thanks for your extensive information. But note that most problems are already solved by now (except for the LED where the SPI is still too fast). I'll add a big README at the start of the repo to show that it is now only there for archive purposes.

For my solution see [this chapter](https://github.com/c-mauderer/openwrt-Ubiquity-Aircube-notes/blob/master/notes/Notes.asciidoc#searching-for-alternative-flashing-methods-part-2) as well as the [OpenWRT Forum](https://forum.openwrt.org/t/add-openwrt-support-for-ubiquity-aircube-isp-ubnt-acb-isp/32999/22) (especially the analysis of the ubntbox binary done by Snap) and the Installation method in the [OpenWRT Wiki](https://openwrt.org/toh/ubiquiti/ubiquiti_aircube_isp#via_oem_web_interface).

Basically the firmware on the AirCube doesn't allow any binaries to access the flash except if they are named `/sbin/fwupdate.real`. That made it necessary to put an overlayfs into place to "overwrite" that binary with one that doesn't check for the signature.

**Comment by andrewhodel** [comment]
I just saw that the openwrt website had poor documentation on this subject,
and saw that you had actually written something useful so I copied and
pasted my private notes to you.



Andrew

**Comment by andrewhodel** [comment]
Also the AC model, you cannot flash it from the web ui with the ISP model
fw.

Also, you can just reflash redboot, if it's got the key built into it then
ignore the fw bin range holding it.

**Comment by c-mauderer** [comment]
> I just saw that the openwrt website had poor documentation on this subject, and saw that you had actually written something useful so I copied and pasted my private notes to you.

That's entirely OK. Do you think that there is something worth adding to the OpenWRT wiki in my notes? I tried to add everything relevant to the AirCube wiki page.

> Also the AC model, you cannot flash it from the web ui with the ISP model fw.

That is most likely correct. I don't have an AC and therefore didn't test it. But there are a few people in the forum who wanted to try it. I just didn't have the time yet to support them.

> Also, you can just reflash redboot, if it's got the key built into it then ignore the fw bin range holding it.

In case of the Aircube it's a U-Boot. I haven't found a solution that would write anything to the flash except for the described one. Even the recovery mode in U-Boot checks for the signature. The only possibility was to open the case to get to the serial U-Boot console. From there it was no problem to write a unsigned binary like described [in "Via Serial Console + U-Boot" in the wiki ](https://forum.openwrt.org/t/add-openwrt-support-for-ubiquity-aircube-isp-ubnt-acb-isp/32999/22).

**Comment by andrewhodel** [comment]
I think those developers are smart enough to let you type in an
email address then get an email with a code to edit the wiki :)

I can mail you one, just send me a PO BOX.

I know it's a lot of work.

**Comment by c-mauderer** [comment]
> I think those developers are smart enough to let you type in an email address then get an email with a code to edit the wiki :) I can mail you one, just send me a PO BOX. I know it's a lot of work.

Sorry, I don't entirely get you here. I already have an account at the OpenWRT wiki. Most of the aircube page is written by me. That's why I asked whether you think I should add anything else. The notes that I did in this repository were mostly some kind of background notes and history for myself. I tried to integrate only the relevant parts into the OpenWRT wiki.

**Comment by c-mauderer** [comment]
PS: If you want to send anything outside of github to me: There should be a mail address in my github profile that is visible for all registered users.

**Comment by andrewhodel** [comment]
Sorry, I jumped ahead and assumed you were aware that I was capable of
posting to the wiki too.  I was just explaining why I didn't.

Non daily contact git style programming work, sorry I'm quite sure about it
all.

**Comment by c-mauderer** [comment]
> Sorry, I jumped ahead and assumed you were aware that I was capable of posting to the wiki too. I was just explaining why I didn't.

I think we just talked past each other.

If it's OK for you I'll close that issue.

**Comment by andrewhodel** [comment]
I guess thinking a temporary po box is possible to click and buy and ship
to online is a dream of the devil there and I just don't worry about giving
out a po box.

I don't think we are talking past each other, I just think there are
acceptances that have been forgotten everywhere that don't really reflect
the technological truth or that of any else's stuck points.

Sure, close it.  Email me andrewhodel@gmail.com if you want me to ship you
an airCube AC I really would like to be able to build that firmware.

**Comment by c-mauderer** [comment]
Ah, sorry. That was a problem that I'm not a native speaker. I somehow assumed that you mean an Email with "PO BOX". Not that used to English names for real hardware messages ;)
I'll answer to your mail.

**Comment by andrewhodel** [comment]
no no!  just like a locker box where you get the key from the front desk
each time.
